smithy-model/src/main/java/software/amazon/smithy/model/loader/AbstractMutableModelFile.java

```diff
@@ -285,10 +260,106 @@ private <S extends Shape, B extends AbstractShapeBuilder<? extends B, S>> Option
                 builder.addTrait(trait);
             }
             return Optional.of(builder.build());
+        } catch (IllegalStateException e) {
+            if (builder.getShapeType() == ShapeType.MEMBER && ((MemberShape.Builder) builder).getTarget() == null) {
+                events.add(ValidationEvent.builder()
+                        .severity(Severity.ERROR)
+                        .id(Validator.MODEL_ERROR)
+                        .shapeId(builder.getId())
+                        .sourceLocation(builder.getSourceLocation())
+                        .message("Member target was elided, but no parent shape contained a matching member name or "
+                                + "resource identifier name.")
+                        .build());
+                return Optional.empty();
+            }
+            throw e;
         } catch (SourceException e) {
             events.add(ValidationEvent.fromSourceException(e, "", builder.getId()));
             resolvedTraits.clearTraitsForShape(builder.getId());
             return Optional.empty();
         }
     }
+
+    private static class ApplyMixin implements PendingShapeModifier {
+        private ShapeId mixin;
+        private final List<ValidationEvent> events;
+
+        ApplyMixin(ShapeId mixin, List<ValidationEvent> events) {
+            this.mixin = mixin;
+            this.events = events;
+        }
+
+        @Override
+        public Set<ShapeId> getDependencies() {
+            return Collections.singleton(mixin);
+        }
+
+        @Override
+        public void modifyMember(
+                MemberShape.Builder memberBuilder,
+                TraitContainer resolvedTraits,
+                Map<ShapeId, Shape> shapeMap
+        ) {
+            // Fast-fail the common case.
+            if (memberBuilder.getTarget() != null) {
+                return;
+            }
+
+            // Members inherited from mixins can have their targets elided, so here we set them
+            // to the target defined in the mixin.
+            Shape mixinShape = shapeMap.get(mixin);
+            String name = memberBuilder.getId().getMember().get();
+            if (mixinShape.getMember(name).isPresent()) {
+                memberBuilder.target(mixinShape.getMember(name).get().getTarget());
+            }
+        }
+
+        @Override
+        public void modifyShape(
+                AbstractShapeBuilder<?, ?> builder,
+                Map<String, MemberShape.Builder> memberBuilders,
+                TraitContainer resolvedTraits,
+                Map<ShapeId, Shape> shapeMap
+        ) {
+            Shape mixinShape = shapeMap.get(mixin);
```

Comment:
Lost the comment here when this was broken out.

Reply:
The comment omitted no longer makes sense. There's javadoc on the `ApplyMixin` class that serves the same purpose.